pkg/optimizer/flatzinc_model.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020 IBM Corp.
+// SPDX-License-Identifier: Apache-2.0
+
+package optimizer
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"strings"
+)
+
+type FlatZincParam struct {
+	NumInstances uint // NumInstances > 1 means an array of params
+	Name         string
+	Type         string
+	Assignment   string
+}
+
+func (fzp FlatZincParam) ParamDeclaration() string {
+	if fzp.NumInstances == 1 {
+		return fmt.Sprintf("%s: %s = %s;\n", fzp.Type, fzp.Name, fzp.Assignment)
+	}
+	return fmt.Sprintf("array [1..%d] of %s: %s = %s;\n", fzp.NumInstances, fzp.Type, fzp.Name, fzp.Assignment)
+}
+
+type Annotations []string
+
+func (annots Annotations) AnnotationString() string {
+	annotsStr := strings.Join(annots, " :: ")
+	if len(annotsStr) > 0 {
+		annotsStr = " :: " + annotsStr
+	}
+	return annotsStr
+}
+
+type FlatZincVariable struct {
+	NumInstances uint // NumInstances > 1 means an array of vars
+	Name         string
+	Type         string
+	Assignment   string
+	Annotations
+}
+
+func (fzv FlatZincVariable) VarDeclaration() string {
+	if fzv.NumInstances == 1 {
+		return fmt.Sprintf("var %s: %s%s;\n", fzv.Type, fzv.Name, fzv.AnnotationString())
+	}
+	return fmt.Sprintf("array [1..%d] of var %s: %s%s;\n", fzv.NumInstances, fzv.Type, fzv.Name, fzv.AnnotationString())
+}
+
+type FlatZincConstraint struct {
+	Identifier  string
+	Expressions []string
+	Annotations
+}
+
+func (cnstr FlatZincConstraint) ConstraintStatement() string {
+	exprs := strings.Join(cnstr.Expressions, ", ")
+	return fmt.Sprintf("constraint %s(%s)%s;\n", cnstr.Identifier, exprs, cnstr.AnnotationString())
+}
+
+type SolveGoal int64
+
+const (
+	Satisfy SolveGoal = iota
+	Minimize
```

Comment:
I believe that the config policy code has constants for minimize and maximize.  Should be put somewhere where it can be used by both the config policies and the optimizer.

Reply:
These are integers. Config policies use their textual representation: lines 77, 79.